lib/rubocop/cop/rails/redundant_presence_validation_on_belongs_to.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # Since Rails 5.0 the default for `belongs_to` is `optional: false`
+      # unless `config.active_record.belongs_to_required_by_default` is
+      # explicitly set to `false`. The presence validator is added
+      # automatically, and explicit presence validation is redundant.
+      #
+      # @example
+      #   # bad
+      #   belongs_to :user
+      #   validates :user, presence: true
+      #
+      #   # bad
+      #   belongs_to :user
+      #   validates :user_id, presence: true
+      #
+      #   # bad
+      #   belongs_to :author, foreign_key: :user_id
+      #   validates :user_id, presence: true
+      #
+      #   # good
+      #   belongs_to :user
+      #
+      #   # good
+      #   belongs_to :author, foreign_key: :user_id
+      #
+      class RedundantPresenceValidationOnBelongsTo < Base
+        include RangeHelp
+        extend AutoCorrector
+        extend TargetRailsVersion
+
+        MSG = 'Remove explicit presence validation for `%<association>s`.'
+        RESTRICT_ON_SEND = %i[validates].freeze
+
+        minimum_target_rails_version 5.0
+
+        def_node_matcher :presence_validation?, <<~PATTERN
+          $(
+            send nil? :validates
+            (sym $_)
+            ...
+            $(hash <$(pair (sym :presence) {true hash}) ...>)
+          )
+        PATTERN
+
+        def_node_matcher :optional_option?, <<~PATTERN
+          {
+            (hash <(pair (sym :optional) true) ...>)
+            (hash <(pair (sym :required) false) ...>)
+          }
+        PATTERN
+
+        def_node_matcher :optional?, <<~PATTERN
+          (send nil? :belongs_to _ ... #optional_option?)
+        PATTERN
+
+        def_node_matcher :class_with_required_belongs_to?, <<~PATTERN
+          (
+            begin
+            <
+              ${
+                (send nil? :belongs_to (sym %key) ...)
```

Comment:
What I'm really struggling with here is this node pattern line and this spec that I've marked as `pending`:
```ruby
      belongs_to :author, foreign_key: :user_id
      validates :author_id, presence: true
```

Problem is that I only want this line to match when there is **no** `foreign_key` option passed.
For `belongs_to` **with** a `foreign_key` option, there's a next node pattern line.

So, I want to match:
 - either `belongs_to` without the option hash
 - either `belongs_to` with option hash that has no `foreign_key` option

Reply:
Ok, I've managed to do this. There's a problem found by other spec example.
It will take a while to shuffle things around to fix it without breaking other ones.